pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,371 @@
+	if err != nil {
+		return fmt.Errorf("failed to associate public ip address %s to instance %s: %w", floatingIP.FloatingIP, instance.Name, err)
+	}
+	return nil
+}
+
+func ensureSecurityGroupRules(openstackClientFactory openstackclient.Factory, bastion *extensionsv1alpha1.Bastion, opt *Options, secGroupID string) error {
+	ethers, err := ingressPermissions(bastion)
+	if err != nil {
+		return err
+	}
+
+	shootSecurityGroups, err := getSecurityGroupId(openstackClientFactory, opt.ShootName)
+	if err != nil {
+		return err
+	}
+
+	if len(shootSecurityGroups) == 0 {
+		return errors.New("shootSecurityGroups must not be empty")
+	}
+
+	// remove ingress allow ssh rules if not exist from bastion yaml
+	ipRangeCiders := getIpRangeCidrs(ethers)
+	ingressSSHRules, err := getRuleByName(openstackClientFactory, ingressAllowSSHResourceName(opt.BastionInstanceName), secGroupID)
+	if openstackclient.IgnoreNotFoundError(err) != nil {
+		return err
+	}
+
+	if len(ingressSSHRules) != 0 {
+		for _, rule := range ingressSSHRules {
+			if !ipRangeCiders.Has(rule.RemoteIPPrefix) {
+				err = deleteRule(openstackClientFactory, rule.ID)
+				if err != nil {
+					return fmt.Errorf("failed to delete security group rule %s-%s", rule.Description, rule.ID)
+				}
+			}
+		}
+	}
+
+	// create ingress rules
+	for _, etherItem := range ethers {
+		rules := []rules.CreateOpts{IngressAllowSSH(opt, etherItem.EtherType, secGroupID, etherItem.CIDR), EgressAllowSSHToWorker(opt, secGroupID, shootSecurityGroups[0].ID)}
+		for _, item := range rules {
+			if err := createSecurityGroupRuleIfNotExist(openstackClientFactory, item); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func createSecurityGroupRuleIfNotExist(openstackClientFactory openstackclient.Factory, createOpts rules.CreateOpts) error {
+	if _, err := createRules(openstackClientFactory, createOpts); err != nil {
+		if _, ok := err.(gophercloud.ErrDefault409); ok {
+			return nil
+		}
+		return fmt.Errorf("failed to create Security Group rule %s: %w", createOpts.Description, err)
+	}
+	logger.Info("Security Group Rule created", "security group rule", createOpts.Description)
+	return nil
+}
+
+func ensureEmptySecurityGroup(openstackClientFactory openstackclient.Factory, opt *Options) (groups.SecGroup, error) {
+	securityGroups, err := getSecurityGroupId(openstackClientFactory, opt.SecurityGroup)
+	if err != nil {
+		return groups.SecGroup{}, err
+	}
+
+	if len(securityGroups) != 0 {
+		return securityGroups[0], nil
+	}
+
+	result, err := createSecurityGroup(openstackClientFactory, groups.CreateOpts{
+		Name:        opt.SecurityGroup,
+		Description: opt.SecurityGroup,
+	})
+	if err != nil {
+		return groups.SecGroup{}, err
+	}
+
+	if len(result.Rules) != 0 {
+		for _, rule := range result.Rules {
+			err = deleteRule(openstackClientFactory, rule.ID)
+			if err != nil {
+				return groups.SecGroup{}, fmt.Errorf("failed to delete default rules: %w", err)
+			}
+		}
+	}
```

Comment:
yes, It could be the case, some global policies created by other SA, I have one concern, do we really need to remove the "defualt" rules（unwanted rules）, these rules may be created for global policy by other Teams ( eg. multi-cloud or CCEE) in the furture for security purpose or compliance from their point of view, etc. then due to SA permission, it may be don't have permission to remove it. I was met in Azure can't remove/update rules directly due to lack of SA permission unless modifying from security group level.  What about we take care of our rules only？in this way, doesn't matter global policy changes or is revoked from there, we can always guarantee our rules to be applied and manage the security access.

Reply:
@petersutter any comments for this keep default rules, the remove following code and then add `EgressDenyAll` back? 

```	
if len(result.Rules) != 0 {
		for _, rule := range result.Rules {
			err = deleteRule(openstackClientFactory, rule.ID)
			if err != nil {
				return groups.SecGroup{}, fmt.Errorf("failed to delete default rules: %w", err)
			}
		}
	}
```


```
// EgressDenyAll egress deny all
func EgressDenyAll(opt *Options, secGroupID string) rules.CreateOpts {
	return rules.CreateOpts{
		Direction:    "egress",
		Description:  egressDenyAllResourceName(opt.BastionInstanceName),
		PortRangeMin: 1,
		PortRangeMax: 65535,
		Protocol:     "tcp",
		EtherType:    rules.EtherType4,
		SecGroupID:   secGroupID,
	}
}
```

Reply:
@tedteng, the difference on azure is, that you can also add rules to DENY, which is not possible on openstack. I doubt that there is a use-case, where for security purposes a rule is added to ALLOW something. This does not make much sense.
Hence, I stick to my initial proposal. Should we really encounter an issue in this case, which I highly doubt, we can make a fix. Until then we remove all rules that ALLOW something that we do not want.

Reply:
@petersutter committed in e5ed224

Reply:
This is what I wrote initially and I still think this is the superior approach than the current implementation:
>ALWAYS reconcile the rules which means:
> 
> * Get all rules of the bastion security group to check for wanted and unwanted rules
> * add wanted rules (if not already there)
> * AFTERWARDS remove unwanted rules (except there is a way to apply all at once)

please see my proposal https://github.com/tedteng/gardener-extension-provider-openstack/pull/1